### ISSUE TITLE
SPMF_NEP with different matrix types

### ISSUE THREAD

**Issue by maxbennedich** [nep-pack/NonlinearEigenproblems.jl#114]
When adding https://github.com/nep-pack/NonlinearEigenproblems.jl/pull/108/commits/2342484a49e1e226b59699a58eae31ae4d847182, I realized that SPMF_NEP can fail in the following part if the matrix types are not all the same:
```julia
if !(eltype(AA) <: SparseMatrixCSC)
    # Dense
    this=SPMF_NEP{typeof(AA[1]),Ftype}(n,AA,fii,Schur_fact,false);
else ...
```
It uses the type of the first matrix to convert all other matrices. This won't always work, for example if the first matrix type is more specific than the others:
```julia
julia> AA = [randn(3,3)', rand(3,3)];

julia> s = SPMF_NEP{typeof(AA[1]),Ftype}(n,AA,fii,false,false)
ERROR: MethodError: Cannot `convert` an object of type Array{Float64,2} to an object of type Adjoint{Float64,Array{Float64,2}}
```
Either we could change this code to use `eltype(AA)` instead of `typeof(AA[1])` as the type (or add a helper method to let the compiler figure it out; we'd need to also make sure that mixed types work in other parts of the code), or extend the commit I did to also enforce that the specific type is the same for all input matrices. Thoughts?

**Comment by jarlebring** [comment]
After checking if there are mixed sparse and dense matrices, we could promote all matrices to a common type determined by `T_mat=mapreduce(typeof,promote_type,AA)`?

**Comment by maxbennedich** [comment]
Yes, that should work. Most of the time, I suspect it'd be the same as `eltype(AA)`, but one could imagine that the container is more generic.

**Comment by maxbennedich** [comment]
Alright, I did the promotion now as part of #108. Should be ready to merge, unless there's anything else.

**Comment by maxbennedich** [comment]
Fixed and merged to master in #108.
